### PR TITLE
Make Avalonia UI the default on all platforms and move in-process

### PR DESCRIPTION
**Pull request by mjcheetham**
Today GCM has an architecture that delegates all GUI prompting to (bundled) external executables, using WPF-based helpers on Windows and Avalonia UI helpers on Mac and Linux. Each host provider has its own GUI helper executable, meaning we currently bundle four (GitHub.UI, GitLab.UI, Atlassian.Bitbucket.UI, git-credential-manager.ui [for generic auth]) helpers in addition to the core git-credential-manager CLI executable.

On Linux we use the "publish as a single-file" model for each of these executables, meaning we have a copy of the CLR included for each executable! That means five CLRs!

Now that the Avalonia UI-based GUI helpers have matured, we can look at dropping the WPF helpers and reduce the effort required to ship new GUI by 50% (we no longer need to do everything twice: once with WPF and once with Avalonia UI).

In addition, in a recent spike, I found that it may be possible to bring the Avalonia UI helpers in-process, with nearly zero net-new code. This would help reduce our installation footprint on Linux down to approximately 1/4 of the current size. The single executable model would also help decrease build times, and increase end-user performance since we no longer pay the process startup costs using external processes!

We need to be mindful that this change may result in a small increase in the Windows installation as we now bring in extra libraries used to support Avalonia UI, whereas previously we relied on WPF libraries as included with Windows 'for free'. This is expected to be minimal.

As for the current GUI extensibility model, we are not aware of any consumers, but it would not be much effort to keep the existing external process model, but now just default to using our in-process GUI options in the absence of external configuration. This still gives tools like VS or VSCode options to integrate native UI for GCM in the future.

Initially, we are still keeping the WPF helpers around on Windows but disabled by default.
We can drop these after at least one release using the in-proc Avalonia helpers.

To switch back to the WPF helpers, use `credential.devUseLegacyUIHelpers` or `GCM_DEV_USELEGACYUIHELPERS`.

### CODE REVIEW

File: src/shared/GitLab/GitLabAuthentication.cs [hunk 1]
```diff
@@ -67,15 +70,69 @@ public async Task<AuthenticationPromptResult> GetAuthenticationAsync(Uri targetU
...
+            {
+                viewModel.UserName = userName;
+                viewModel.TokenUserName = userName;
+            }
+
+            await AvaloniaUi.ShowViewAsync<CredentialsView>(viewModel, GetParentWindowHandle(), CancellationToken.None);
+
+            if (!viewModel.WindowResult)
+            {
+                throw new Trace2Exception(Context.Trace2, "User cancelled dialog.");
```
Comment (ldennington) [inline src/shared/GitLab/GitLabAuthentication.cs:111]: Is there a reason we can't use `ThrowIfWindowCancelled()` here?

File: src/shared/Core/Core.csproj [hunk 2]
```diff
@@ -21,12 +21,13 @@
     <PackageReference Include="Microsoft.Identity.Client" Version="4.52.0" />
     <PackageReference Include="Microsoft.Identity.Client.Extensions.Msal" Version="2.28.0" />
     <PackageReference Include="System.CommandLine" Version="2.0.0-beta1.21216.1" />
-    <PackageReference Include="Avalonia" Version="0.10.18" />
-    <PackageReference Include="Avalonia.Desktop" Version="0.10.18" />
+    <PackageReference Include="Avalonia" Version="11.0.0-preview6" />
```
Comment (ldennington) [inline src/shared/Core/Core.csproj:29]: Wow! Big jump.

**Review by ldennington — APPROVED** [review]
This is so exciting! Great work, Matthew! 🚀

File: src/shared/Core/Constants.cs [hunk 3]
```diff
@@ -102,6 +102,7 @@ public static class EnvironmentVariables
             public const string OAuthDeviceEndpoint      = "GCM_OAUTH_DEVICE_ENDPOINT";
             public const string OAuthClientAuthHeader    = "GCM_OAUTH_USE_CLIENT_AUTH_HEADER";
             public const string OAuthDefaultUserName     = "GCM_OAUTH_DEFAULT_USERNAME";
+            public const string GcmDevUseLegacyUiHelpers = "GCM_DEV_USELEGACYUIHELPERS";
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply src/shared/Core/Constants.cs:106]: This isn't something we want to necessarily advertise or be something that users can rely on being there for the long term. It's a 'hidden' setting – we don't document it.

File: src/shared/Core/Authentication/OAuthAuthentication.cs [hunk 4]
```diff
@@ -60,74 +60,83 @@ public OAuthAuthentication(ICommandContext context)
...
+                    if ((modes & OAuthAuthenticationModes.DeviceCode) != 0) deviceItem = menu.Add("Device code");
 
-                    default:
-                        throw new Trace2Exception(Context.Trace2,
-                            $"Unknown mode value in response '{responseMode}'");
-                }
+                    // Default to the 'first' choice in the menu
+                    TerminalMenuItem choice = menu.Show(0);
+
+                    if (choice == browserItem) goto case OAuthAuthenticationModes.Browser;
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply src/shared/Core/Authentication/OAuthAuthentication.cs:129]: Yeah, the C# `goto case` keywords used together is a neat way to achieve this sort of control flow, without allowing for spaghetti jumps (can only jump to another `case`).

File: src/linux/Packaging.Linux/layout.sh [hunk 5]
```diff
@@ -81,15 +80,6 @@ $DOTNET_ROOT/dotnet publish "$GCM_SRC" \
 	-p:PublishSingleFile=true \
 	--output="$(make_absolute "$PAYLOAD")" || exit 1
 
-echo "Publishing core UI helper..."
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply src/linux/Packaging.Linux/layout.sh:84]: Builds are much faster and will be easier to maintain or modify going forwards (at least once we can drop the WPF helpers entirely).

File: src/shared/GitLab/GitLabAuthentication.cs [hunk 6]
```diff
@@ -70,131 +70,141 @@ public async Task<AuthenticationPromptResult> GetAuthenticationAsync(Uri targetU
...
+
+            return GetAuthenticationViaTty(targetUri, userName, modes);
+        }
+
+        private AuthenticationPromptResult GetAuthenticationViaTty(Uri targetUri, string userName, AuthenticationModes modes)
+        {
+            switch (modes)
+            {
+                case AuthenticationModes.Basic:
+                    ThrowIfUserInteractionDisabled();
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply src/shared/GitLab/GitLabAuthentication.cs:141]: `Bitbucket` and `GitHubAuthentication` both make calls to `ThrowIfUserInteractionDisabled` and `ThrowIfTerminalPromptsDisabled`, but they do so in the main/public methods, not these `ViaTty/Ui/Helper` methods.

Really there's no reason why the GitLab implementation shouldn't also simplify and move all these checks to higher in the call stack, but this is just a refactoring (it looks like this in the existing code today).

[the remark this answers is not in the document]
Comment (ldennington) [reply src/shared/GitLab/GitLabAuthentication.cs:141]: Makes sense!

File: src/shared/Core/Core.csproj (hunk 2 above)
Comment (mjcheetham) [reply src/shared/Core/Core.csproj:29]: Yeah they moved from `0.10` to dropping the `0.` prefix with v11

File: src/shared/GitLab/GitLabAuthentication.cs (hunk 1 above)
Comment (mjcheetham) [reply src/shared/GitLab/GitLabAuthentication.cs:111]: Good point! Updated.

File: src/shared/GitLab/GitLabAuthentication.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mjcheetham) [reply src/shared/GitLab/GitLabAuthentication.cs]: Fixed this by adding another commit that pulls these checks up.

File: src/shared/Core/UI/ViewModels/WindowViewModel.cs [hunk 7]
```diff
@@ -35,13 +35,20 @@ public bool ShowCustomChrome
             get => ShowCustomChromeOverride || (ExtendClientArea && !PlatformUtils.IsMacOS());
         }
 
+        public bool ShowCustomWindowBorder
+        {
+            // Draw the window border explicitly on Windows
+            get => ShowCustomChrome && PlatformUtils.IsWindows();
```
[the remark this answers is not in the document]
Comment (mjcheetham) [reply src/shared/Core/UI/ViewModels/WindowViewModel.cs:41]: Oh and on macOS the thin 1-pixel window border is always drawn, but on Windows it's not.
This means on Windows we want to manually draw a 1-px border or else our window can blend in to the background.

**Pull request merged**